### PR TITLE
Add new allowLeadingWhitespace and allowTrailingWhitespace flags

### LINKED ISSUE #89: Add rule for leading/trailing whitespace
Raised at https://gerrit.wikimedia.org/r/c/mediawiki/core/+/657539

### PR DESCRIPTION
**Pull request by jdforrester**
Initially defaulted to true; given almost exclusive user is MediaWiki, we
should probably default them to false, but that would be a breaking change
so want to make sure this works and meets the needs of the user that
requested this first.

Fixes #89

### CODE REVIEW

File: src/banana.js [hunk 1]
```diff
@@ -32,16 +35,17 @@ module.exports = function bananaChecker( dir, options, logErr ) {
...
 		requireMetadata: true,
 
+		allowLeadingWhitespace: true,
+		allowTrailingWhitespace: true,
+
 		skipIncompleteMessageDocumentation: []
 	}, options );
 
 	const jsonFilenameRegex = /(.*)\.json$/;
+	const leadingWhitespaceRegex = /^\s+/;
```
Comment (edg2s) [inline src/banana.js:45]: Micro-optimisation: The + in unnecessary. Detecting one character of whitespace is sufficient.

File: src/banana.js [hunk 2]
```diff
@@ -180,6 +204,40 @@ module.exports = function bananaChecker( dir, options, logErr ) {
 		} );
 	}
 
+	let sourceMessageLeadingWhitespace = [];
+	let documentationMessageLeadingWhitespace = [];
+	if ( !options.allowLeadingWhitespace ) {
+		sourceMessageLeadingWhitespace = sourceMessageKeys.filter(
```
[the remark this answers is not in the document]
Comment (Krinkle) [reply src/banana.js:210]: Afaik most rules, including this one, report which message keys violate the rule. sourceMessageLeadingWhitespace is iterated lower down.

[the remark this answers is not in the document]
Comment (edg2s) [reply src/banana.js:210]: 👍

File: src/banana.js [hunk 3]
```diff
@@ -180,6 +204,40 @@ module.exports = function bananaChecker( dir, options, logErr ) {
...
+			function ( message ) {
+				return sourceMessages[ message ].match( leadingWhitespaceRegex );
```
Comment (edg2s) [inline src/banana.js:212]: Prefer regex.test() which returns a boolean.

Comment (jdforrester) [reply src/banana.js:212]: We consistently use .match() in this codebase currently. I'd rather not make that change just here, if that's OK?

File: src/banana.js [hunk 4]
```diff
@@ -180,6 +192,40 @@ module.exports = function bananaChecker( dir, options, logErr ) {
...
+		);
+
+		documentationMessageTrailingWhitespace = documentationMessageKeys.filter(
+			function ( message ) {
+				return trailingWhitespaceRegex.test( documentationMessages[ message ] );
+			}
+		);
+	}
+
+	let sourceMessageMissing = [];
```
Comment (edg2s) [inline src/banana.js:227]: Linter is complaining this has already been declared.

Comment (jdforrester) [reply src/banana.js:227]: How odd. It passes locally?

Comment (jdforrester) [reply src/banana.js:239]: Oh, it pushed the wrong HEAD. Meh. Fixed.

**Review by edg2s — APPROVED** [review]

**Pull request merged**